File: Wed_Thời Trang Nam/ThuanBackend/Thuan.API/Controllers/UploadController.cs
// Controllers/UploadController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Thuan.API.Data;
using Thuan.API.Models;
using System.IO;

namespace Thuan.API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class UploadController : ControllerBase
    {
        private readonly AppDbContext _context;
        private readonly IWebHostEnvironment _webHostEnvironment;

        // Đường dẫn upload (tùy chỉnh)
        private readonly string _uploadPath = "uploads";

        public UploadController(AppDbContext context, IWebHostEnvironment webHostEnvironment)
        {
            _context = context;
            _webHostEnvironment = webHostEnvironment;
        }

        private string GetUploadsDirectory()
        {
            var webRoot = _webHostEnvironment.WebRootPath;
            if (string.IsNullOrEmpty(webRoot))
                webRoot = Path.Combine(_webHostEnvironment.ContentRootPath, "wwwroot");
            var dir = Path.Combine(webRoot, _uploadPath);
            if (!Directory.Exists(dir))
                Directory.CreateDirectory(dir);
            return dir;
        }

        private string ToPublicUrl(string relativePath)
        {
            var req = HttpContext.Request;
            return $"{req.Scheme}://{req.Host}{relativePath}";
        }

        // POST /api/upload/multiple - Upload nhiều file (ảnh/video sản phẩm)
        [HttpPost("multiple")]
        [AllowAnonymous]
        public async Task<IActionResult> UploadMultiple([FromForm] List<IFormFile> files, [FromForm] string? description)
        {
            if (files == null || files.Count == 0)
                return BadRequest(new { message = "Chưa chọn file" });

            var results = new List<object>();
            foreach (var file in files)
            {
                var single = await SaveUploadedFile(file, null, description);
                if (single.Error != null)
                    return BadRequest(new { message = single.Error });
                results.Add(single.Result!);
            }
            return Ok(new { message = $"Upload {results.Count} file thành công", data = results });
        }

        // POST /api/upload - Upload file
        [HttpPost]
        [AllowAnonymous]
        public async Task<IActionResult> UploadFile([FromForm] IFormFile file, [FromForm] int? userId, [FromForm] string? description)
        {
            if (file == null || file.Length == 0)
                return BadRequest(new { message = "File không được để trống" });

            // Validate file size (max 100MB)
            const long maxFileSize = 100 * 1024 * 1024;
            if (file.Length > maxFileSize)
                return BadRequest(new { message = "File quá lớn (max 100MB)" });

            // Validate file type
            string[] allowedExtensions = { ".jpg", ".jpeg", ".png", ".gif", ".mp4", ".webm", ".avi", ".mov", ".pdf", ".doc", ".docx" };
            var fileExtension = Path.GetExtension(file.FileName).ToLowerInvariant();

            if (!allowedExtensions.Contains(fileExtension))
                return BadRequest(new { message = "Định dạng file không được phép" });

            var saved = await SaveUploadedFile(file, userId, description);
            if (saved.Error != null)
                return StatusCode(500, new { message = saved.Error });
            return Ok(saved.Result);
        }

        private async Task<(object? Result, string? Error)> SaveUploadedFile(IFormFile file, int? userId, string? description)
        {
            if (file == null || file.Length == 0)
                return (null, "File không được để trống");

            const long maxFileSize = 100 * 1024 * 1024;
            if (file.Length > maxFileSize)
                return (null, "File quá lớn (max 100MB)");

            string[] allowedExtensions = { ".jpg", ".jpeg", ".png", ".gif", ".webp", ".mp4", ".webm", ".avi", ".mov", ".pdf", ".doc", ".docx" };
            var fileExtension = Path.GetExtension(file.FileName).ToLowerInvariant();
            if (!allowedExtensions.Contains(fileExtension))
                return (null, "Định dạng file không được phép");

            try
            {
                var uploadsDirectory = GetUploadsDirectory();
                string uniqueFileName = Guid.NewGuid().ToString() + fileExtension;
                var filePath = Path.Combine(uploadsDirectory, uniqueFileName);

                using (var fileStream = new FileStream(filePath, FileMode.Create))
                    await file.CopyToAsync(fileStream);

                string fileType = GetFileType(fileExtension);
                var relativeUrl = $"/{_uploadPath}/{uniqueFileName}";

                var uploadedFile = new UploadedFile
                {
                    FileName = file.FileName,
                    FileUrl = relativeUrl,
                    FileType = fileType,
                    FileSizeBytes = file.Length,
                    UserId = userId,
                    Description = description,
                    UploadedAt = DateTime.UtcNow
                };

                _context.UploadedFiles.Add(uploadedFile);
                await _context.SaveChangesAsync();

                return (new
                {
                    message = "Upload thành công",
                    id = uploadedFile.Id,
                    fileName = uploadedFile.FileName,
                    fileUrl = relativeUrl,
                    publicUrl = ToPublicUrl(relativeUrl),
                    fileType = uploadedFile.FileType,
                    fileSizeBytes = uploadedFile.FileSizeBytes
                }, null);
            }
            catch (Exception ex)
            {
                return (null, $"Lỗi upload: {ex.Message}");
            }
        }

        // GET /api/upload/{id}/download - Download file
        [HttpGet("{id}/download")]
        public async Task<IActionResult> DownloadFile(int id)
        {
            var uploadedFile = await _context.UploadedFiles.FindAsync(id);
            if (uploadedFile == null)
                return NotFound(new { message = "File không tìm thấy" });

            var filePath = Path.Combine(_webHostEnvironment.WebRootPath, uploadedFile.FileUrl.TrimStart('/'));
            if (!System.IO.File.Exists(filePath))
                return NotFound(new { message = "File không tồn tại trên server" });

            // Tăng download count
            uploadedFile.DownloadCount++;
            _context.UploadedFiles.Update(uploadedFile);
            await _context.SaveChangesAsync();

            var fileBytes = System.IO.File.ReadAllBytes(filePath);
            return File(fileBytes, GetContentType(Path.GetExtension(filePath)), uploadedFile.FileName);
        }

        // GET /api/upload - Lấy danh sách file
        [HttpGet]
        public async Task<IActionResult> GetUploadedFiles(int? userId = null, string? fileType = null)
        {
            IQueryable<UploadedFile> query = _context.UploadedFiles;

            if (userId.HasValue)
                query = query.Where(f => f.UserId == userId);

            if (!string.IsNullOrEmpty(fileType))
                query = query.Where(f => f.FileType == fileType);

            var files = await query
                .OrderByDescending(f => f.UploadedAt)
                .Select(f => new
                {
                    f.Id,
                    f.FileName,
                    f.FileUrl,
                    f.FileType,
                    f.FileSizeBytes,
                    f.UserId,
                    f.Description,
                    f.UploadedAt,
                    f.DownloadCount
                })
                .ToListAsync();

            return Ok(new { data = files, count = files.Count });
        }

        // DELETE /api/upload/{id} - Xóa file
        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteFile(int id)
        {
            var uploadedFile = await _context.UploadedFiles.FindAsync(id);
            if (uploadedFile == null)
                return NotFound(new { message = "File không tìm thấy" });

            try
            {
                // Xóa file từ disk
                var filePath = Path.Combine(_webHostEnvironment.WebRootPath, uploadedFile.FileUrl.TrimStart('/'));
                if (System.IO.File.Exists(filePath))
                    System.IO.File.Delete(filePath);

                // Xóa khỏi database
                _context.UploadedFiles.Remove(uploadedFile);
                await _context.SaveChangesAsync();

                return Ok(new { message = "Xóa file thành công" });
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { message = $"Lỗi xóa: {ex.Message}" });
            }
        }

        // === HỖ TRỢ ===
        private string GetFileType(string extension)
        {
            return extension.ToLowerInvariant() switch
            {
                ".mp4" or ".webm" or ".avi" or ".mov" => "video",
                ".jpg" or ".jpeg" or ".png" or ".gif" => "image",
                ".pdf" or ".doc" or ".docx" => "document",
                _ => "other"
            };
        }

        private string GetContentType(string extension)
        {
            return extension.ToLowerInvariant() switch
            {
                ".mp4" => "video/mp4",
                ".webm" => "video/webm",
                ".avi" => "video/avi",
                ".mov" => "video/quicktime",
                ".jpg" or ".jpeg" => "image/jpeg",
                ".png" => "image/png",
                ".gif" => "image/gif",
                ".pdf" => "application/pdf",
                ".doc" => "application/msword",
                ".docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                _ => "application/octet-stream"
            };
        }
    }
}

File: Wed_Thời Trang Nam/ThuanBackendNode/controllers/UploadController.js
// controllers/UploadController.js
const fs = require('fs');
const path = require('path');
const multer = require('multer');
const { getMySQLPool } = require('../config/db');

// Ensure upload directory exists
const uploadDir = path.join(__dirname, '..', 'uploads');
if (!fs.existsSync(uploadDir)) {
    fs.mkdirSync(uploadDir, { recursive: true });
}

// Multer configuration
const storage = multer.diskStorage({
    destination: (req, file, cb) => {
        cb(null, uploadDir);
    },
    filename: (req, file, cb) => {
        const ext = path.extname(file.originalname).toLowerCase();
        const uniqueName = `${Date.now()}-${Math.round(Math.random() * 1E9)}${ext}`;
        cb(null, uniqueName);
    }
});

// Max size 100MB
const upload = multer({
    storage: storage,
    limits: { fileSize: 100 * 1024 * 1024 }
});

const getFileType = (ext) => {
    switch (ext) {
        case '.mp4':
        case '.webm':
        case '.avi':
        case '.mov':
            return 'video';
        case '.jpg':
        case '.jpeg':
        case '.png':
        case '.gif':
        case '.webp':
            return 'image';
        case '.pdf':
        case '.doc':
        case '.docx':
            return 'document';
        default:
            return 'other';
    }
};

const getContentType = (ext) => {
    switch (ext) {
        case '.mp4': return 'video/mp4';
        case '.webm': return 'video/webm';
        case '.avi': return 'video/avi';
        case '.mov': return 'video/quicktime';
        case '.jpg':
        case '.jpeg': return 'image/jpeg';
        case '.png': return 'image/png';
        case '.gif': return 'image/gif';
        case '.pdf': return 'application/pdf';
        case '.doc': return 'application/msword';
        case '.docx': return 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
        default: return 'application/octet-stream';
    }
};

// Helper function to save file details into MySQL db
async function saveUploadedFile(file, userId, description, req) {
    const ext = path.extname(file.originalname).toLowerCase();
    const fileType = getFileType(ext);
    const relativeUrl = `/uploads/${file.filename}`;

    const pool = getMySQLPool();
    const [result] = await pool.query(
        `INSERT INTO UploadedFiles (file_name, file_url, file_type, file_size_bytes, user_id, description, uploaded_at, download_count)
         VALUES (?, ?, ?, ?, ?, ?, NOW(), 0)`,
        [file.originalname, relativeUrl, fileType, file.size, userId || null, description || '']
    );

    const publicUrl = `${req.protocol}://${req.get('host')}${relativeUrl}`;

    return {
        message: 'Upload thành công',
        id: result.insertId,
        fileName: file.originalname,
        fileUrl: relativeUrl,
        publicUrl: publicUrl,
        fileType: fileType,
        fileSizeBytes: file.size
    };
}

// POST /api/upload - Single file upload
exports.uploadSingle = [
    upload.single('file'),
    async (req, res) => {
        if (!req.file) {
            return res.status(400).json({ message: 'File không được để trống' });
        }

        const { userId, description } = req.body;

        try {
            const result = await saveUploadedFile(req.file, userId, description, req);
            return res.status(200).json(result);
        } catch (error) {
            console.error('Single upload database error:', error);
            // Cleanup uploaded file on error
            try {
                fs.unlinkSync(req.file.path);
            } catch (err) {}
            return res.status(500).json({ message: `Lỗi upload: ${error.message}` });
        }
    }
];

// POST /api/upload/multiple - Multiple files upload
exports.uploadMultiple = [
    upload.array('files'),
    async (req, res) => {
        if (!req.files || req.files.length === 0) {
            return res.status(400).json({ message: 'Chưa chọn file' });
        }

        const { description } = req.body;
        const results = [];

        try {
            for (const file of req.files) {
                const singleResult = await saveUploadedFile(file, null, description, req);
                results.push(singleResult);
            }
            return res.status(200).json({
                message: `Upload ${results.length} file thành công`,
                data: results
            });
        } catch (error) {
            console.error('Multiple upload database error:', error);
            // Cleanup files
            if (req.files) {
                req.files.forEach(f => {
                    try {
                        fs.unlinkSync(f.path);
                    } catch (err) {}
                });
            }
            return res.status(500).json({ message: `Lỗi upload: ${error.message}` });
        }
    }
];

// GET /api/upload/:id/download - Download file
exports.downloadFile = async (req, res) => {
    const { id } = req.params;

    try {
        const pool = getMySQLPool();
        const [files] = await pool.query('SELECT * FROM UploadedFiles WHERE id = ?', [id]);

        if (files.length === 0) {
            return res.status(404).json({ message: 'File không tìm thấy' });
        }

        const file = files[0];
        const filePath = path.join(__dirname, '..', file.file_url);

        if (!fs.existsSync(filePath)) {
            return res.status(404).json({ message: 'File không tồn tại trên server' });
        }

        // Increment download count
        await pool.query('UPDATE UploadedFiles SET download_count = download_count + 1 WHERE id = ?', [id]);

        const ext = path.extname(file.file_name).toLowerCase();
        res.setHeader('Content-Type', getContentType(ext));
        res.setHeader('Content-Disposition', `attachment; filename="${encodeURIComponent(file.file_name)}"`);
        
        return res.sendFile(filePath);
    } catch (error) {
        console.error('Download file error:', error);
        return res.status(500).json({ message: `Lỗi tải file: ${error.message}` });
    }
};

// GET /api/upload - Get all uploaded files list
exports.getUploadedFiles = async (req, res) => {
    const { userId, fileType } = req.query;

    try {
        const pool = getMySQLPool();
        let query = 'SELECT id, file_name as fileName, file_url as fileUrl, file_type as fileType, file_size_bytes as fileSizeBytes, user_id as userId, description, uploaded_at as uploadedAt, download_count as downloadCount FROM UploadedFiles WHERE 1=1';
        const params = [];

        if (userId) {
            query += ' AND user_id = ?';
            params.push(userId);
        }

        if (fileType) {
            query += ' AND file_type = ?';
            params.push(fileType);
        }

        query += ' ORDER BY uploaded_at DESC';

        const [files] = await pool.query(query, params);
        return res.status(200).json({ data: files, count: files.length });
    } catch (error) {
        console.error('Get uploaded files error:', error);
        return res.status(500).json({ message: `Lỗi lấy danh sách file: ${error.message}` });
    }
};

// DELETE /api/upload/:id - Delete file
exports.deleteFile = async (req, res) => {
    const { id } = req.params;

    try {
        const pool = getMySQLPool();
        const [files] = await pool.query('SELECT * FROM UploadedFiles WHERE id = ?', [id]);

        if (files.length === 0) {
            return res.status(404).json({ message: 'File không tìm thấy' });
        }

        const file = files[0];
        const filePath = path.join(__dirname, '..', file.file_url);

        // Delete from disk
        if (fs.existsSync(filePath)) {
            try {
                fs.unlinkSync(filePath);
            } catch (err) {
                console.warn(`Could not delete physical file at ${filePath}:`, err);
            }
        }

        // Delete from DB
        await pool.query('DELETE FROM UploadedFiles WHERE id = ?', [id]);

        return res.status(200).json({ message: 'Xóa file thành công' });
    } catch (error) {
        console.error('Delete file error:', error);
        return res.status(500).json({ message: `Lỗi xóa file: ${error.message}` });
    }
};
